### PR TITLE
build: refactor ambient node & jasmine types so they are only included where needed

### PR DESCRIPTION
**Pull request by gregmagolan**
Replaces https://github.com/angular/angular/pull/25471.

This solves a downstream build issue with Material 2 getting node types includes in its build from `angular/packages/types.d.ts`.

### CODE REVIEW

**Comment by gregmagolan** [comment]
Packages that depend on node types: language-service, platform-server, compiler/testing, benchpress, compiler-cli/src/ngtsc/util, tools/symbol-extractor

**Comment by gregmagolan** [comment]
`/packages/animations/browser/src/render/shared.ts` declares `declare const process: any;` to avoid needing node types.

And any ts_library or ng_module that is set to `testonly=1` will use `tsconfig-test.json` which brings in node types via `types = ["node"]`.

**Comment by gregmagolan** [comment]
@alexeagle I looked at removing the `types-node.ts` and `types-node.d.ts` files. There are two issues.

1) You need at least one `.ts` file for `ts_library` since we don't have a `ts_declaration` rule.

2) The sources for ts_library need to be in the same package as the rule so

```
ts_library(
    name = "types-node",
    srcs = ["@angular_deps//:node_modules/@types/node/index.d.ts"],
)
```

doesn't work and errors out with:

```
	File "/private/var/tmp/_bazel_greg/6d4db7eb5dc95ad27238d5cd589d6e42/external/build_bazel_rules_typescript/internal/common/compilation.bzl", line 176, in compile_ts
		fail(("Sources must be in the same pa...))), ...")
attribute srcs: Sources must be in the same package as the ts_library rule, but @angular_deps//:node_modules/@types/node/index.d.ts is not in packages
```

**Comment by gregmagolan** [comment]
`compiles_ts` in rules_typescript `compilation.bzl` does have an `is_library` param which I assume is used for `ts_declaration` in google3

**Comment by alexeagle** [comment]
We have had discussion about ts_declaration and we don't have a good justification for why users should have to distinguish. Meaning, if you always use `ts_library` we can still do the right thing (like check if all `srcs` have extension `.d.ts` or something)
If we want to make it more elegant, we should maybe let ts_library have no `.ts` files.
As for the package, let's do something in the fine-grained deps proposal - when we write a BUILD file for `@types/foo` we could create the `ts_library` for importing the typings

File: packages/BUILD.bazel [hunk 1]
```diff
@@ -2,14 +2,26 @@ package(default_visibility = ["//visibility:public"])
...
 ts_library(
     name = "types",
-    srcs = glob(["*.ts"]),
+    srcs = glob(
+        ["*.ts"],
+        exclude = ["types-node.*"],
+    ),
+)
+
+ts_library(
```
Comment (alexeagle) [inline packages/BUILD.bazel]: leave a TODO for me to revisit this once we have fine-grained deps generating a `node_modules/@types/node/BUILD` file

File: packages/animations/browser/src/render/shared.ts [hunk 2]
```diff
@@ -10,6 +10,10 @@ import {AUTO_STYLE, AnimationEvent, AnimationPlayer, NoopAnimationPlayer, ɵAnim
 import {AnimationStyleNormalizer} from '../../src/dsl/style_normalization/animation_style_normalizer';
 import {AnimationDriver} from '../../src/render/animation_driver';
 
+// We don't include ambient node types in the build
+// so `process` is just declared locally here
```
Comment (alexeagle) [inline packages/animations/browser/src/render/shared.ts]: maybe add to this comment that animations/browser is meant to target browser runtime so technically it shouldn't depend on node

File: packages/bazel/test/ng_package/BUILD.bazel [hunk 3]
```diff
@@ -9,7 +9,7 @@ exports_files(["package.json"])
 
 ts_library(
     name = "core_spec_lib",
-    testonly = True,
+    testonly = 1,
```
Comment (alexeagle) [inline packages/bazel/test/ng_package/BUILD.bazel]: why are these changed? did the formatter do it?
I think `True` is preferred

File: packages/tsconfig-test.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Root tsconfig file for use in all tests.
+ */
+{
+  "extends": "./tsconfig-build.json",
+  "compilerOptions": {
+    "types": ["node"],
```
Comment (alexeagle) [inline packages/tsconfig-test.json]: maybe jasmine should be here too? does the production code currently compile with jasmine ambient types used? (it shouldn't of course)

File: packages/types.d.ts [hunk 5]
```diff
@@ -10,12 +10,18 @@
 
 /// <reference types="hammerjs" />
 /// <reference types="jasmine" />
```
Comment (alexeagle) [inline packages/types.d.ts]: yeah can this move at the same time?

**Review by alexeagle — APPROVED** [review]

File: packages/tsconfig-test.json (hunk 4 above)
Comment (IgorMinar) [reply packages/tsconfig-test.json]: +1 for including jasmine in here.

File: packages/animations/browser/src/render/shared.ts (hunk 2 above)
Comment (gregmagolan) [reply packages/animations/browser/src/render/shared.ts]: 👍

File: packages/types-node.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+// This file intentionally left blank.
```
Comment (IgorMinar) [inline packages/types-node.ts]: please add a note explaining why we need this stub file because when we look at this file in 6 months nobody will remember why it's here - you can copy the PR comment you left explaining this. thanks!

**Comment by IgorMinar** [comment]
@gregmagolan can you please clarify your comment:

> having reference types="node" in the entry points doesn't work anyway since the types are only available for that .ts file and not for all the .ts files in the compilation unit.

I don't think that this is true. Or at least it's not true outside of bazel and running under bazel should not change the visibility of these types.

**Comment by gregmagolan** [comment]
> I don't think that this is true. Or at least it's not true outside of bazel and running under bazel should not change the visibility of these types.

You're right about that. It does work with the `reference types` in the entry point. I read the build error I got wrong there when building a separate ng_module within compiler-ci. This means that `//package:node-types` could be replaced with `reference types` at all the appropriate entry points if that is preferrable. Similarly, `@angular/core/testing` and `@angular/platform-browser/testing` need the jasmine types which could go in the entry points to those modules.

File: packages/compiler-cli/src/ngtsc/util/src/path.ts [hunk 7]
```diff
@@ -6,6 +6,8 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+/// <reference types="node" />
```
Comment (gregmagolan) [inline packages/compiler-cli/src/ngtsc/util/src/path.ts]: there is no `packages/compiler-cli/src/ngtsc/util/index.ts` file although it looks like there should be from the BUILD file:
```
ts_library(
    name = "util",
    srcs = glob([
        "index.ts",
        "src/**/*.ts",
    ]),
    module_name = "@angular/compiler-cli/src/ngtsc/util",
    deps = [
        "//packages:types",
    ],
)
```

File: packages/language-service/language-service.ts [hunk 8]
```diff
@@ -6,6 +6,8 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+/// <reference types="node" />
+
```
Comment (gregmagolan) [inline packages/language-service/language-service.ts]: no `public_api.ts` file in `packages/language-service`

**Review by IgorMinar — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/util/src/path.ts (hunk 7 above)
Comment (alexeagle) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: yeah the "entry point" is in the compiler-cli/ directory, but since this is a dependency of that, I see why you also need to declare the node dependency here

File: packages/private/testing/BUILD.bazel [hunk 9]
```diff
@@ -6,6 +6,7 @@ load("//tools:defaults.bzl", "ng_module")
 
 ng_module(
     name = "testing",
+    testonly = 1,
```
Comment (alexeagle) [inline packages/private/testing/BUILD.bazel]: nit: `True`

**Review by alexeagle — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/util/src/path.ts (hunk 7 above)
Comment (gregmagolan) [reply packages/compiler-cli/src/ngtsc/util/src/path.ts]: in this case it looks like someone forgot to add the `index.ts` file. The ts_library above is from `/packages/compiler-cli/src/ngtsc/util/BUILD.bazel` but there is no `index.ts` file which is in the src glob. Other folder at the same level such as `/packages/compiler-cli/src/ngtsc/factories` have an `index.ts`.

**Comment by gregmagolan** [comment]
Blocked by an aio dgeni build failure after adding the `/// <reference types="x"` to some of the angular source files:

```
$ dgeni ./tools/transforms/angular.io-package
info:    running processor: readFilesProcessor
info:    running processor: extractJSDocCommentsProcessor
info:    running processor: collectExamples
info:    running processor: readTypeScriptModules
error:   Error: No SourceFile found with path node_modules/@types/jasmine/index.d.ts
    at CustomCompilerHost.getSourceFile (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/typescript/services/TsParser/CustomCompilerHost.js:60:15)
    at findSourceFile (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:75050:29)
    at args (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:74987:85)
    at getSourceFileFromReferenceWorker (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:74960:34)
    at processSourceFile (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:74987:13)
    at processTypeReferenceDirective (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:75155:25)
    at processTypeReferenceDirectives (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:75124:17)
    at findSourceFile (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:75095:21)
    at processImportedModules (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:75222:25)
    at findSourceFile (/Users/greg/google/gregmagolan/angular-2/aio/node_modules/dgeni-packages/node_modules/typescript/lib/typescript.js:75098:17)
```

**Comment by jasonaden** [comment]
@alexeagle Can you help get this ready for merge, then put back the `PR action: merge` label?

**Comment by IgorMinar** [comment]
I kicked off a g3 presubmit for this change. [results](http://test/OCL:209014812:BASE:209014880:1534443262456:18c3387a)

**Comment by IgorMinar** [comment]
presubmit is green

**Comment by petebacondarwin** [comment]
How did this pass CircleCI tests?
Now all `aio_preview` jobs are broken because in that job the top level `yarn install` has not been executed, which results in the `node_modules/@types/jasmine/index.d.ts` file not existing.

**Comment by gregmagolan** [comment]
@petebacondarwin That's a good question. Travis was all green if I remember correctly. Is the failure you're talking about the same as I'm seeing on 6.1.x with the same changes? https://github.com/angular/angular/pull/25528

**Comment by petebacondarwin** [comment]
No that error is because we changed the preview server. I forgot to merge the change to the patch branch.

**Comment by petebacondarwin** [comment]
I understand how this happened. The PR was green against the old master, which used Travis to deploy the PR previews. Between that going green and merging it into master, we updating master to use CircleCI for PR previews, which would have caused this PR to fail, if it had been rebased and repushed before merging.  Just one of those things... I don't think it is feasible to insist on PRs being green against the "current" master before merging as it takes so long for Travis and CircleCI to run.
